### PR TITLE
Added support for OSL array parameters.

### LINKED ISSUE #1061: Add support for OSL array parameters.

### PR DESCRIPTION
**Pull request by est77**
Added support for OSL float, vector, point, normal and color array params in projects.
Syntax is "float[] 1.0 2.0 3.0 4.0 ...", "vector[] 1.0 2.0 3.0 1.0 2.0 3.0 ..."
For vectors, colors, points and normals, the number of floats after the type[], must be a multiple of 3.

Closes #1061

### CODE REVIEW

File: src/appleseed/renderer/meta/tests/test_shaderparamparser.cpp [hunk 1]
```diff
@@ -72,6 +72,24 @@ TEST_SUITE(Renderer_Modeling_ShaderParamParser)
         EXPECT_EQ("test_string", parser.parse_string_value());
     }
 
+    TEST_CASE(ShaderParamParserFloatArray)
+    {
+        ShaderParamParser parser("float[] 1.0 2.0 3.0 4.0 5.0 6.0 7.0");
+        EXPECT_EQ(OSLParamTypeFloatArray, parser.param_type());
+        std::vector<float> values;
+        parser.parse_float_array(values);
+        EXPECT_EQ(values.size(), 7);
```
Comment (dictoon) [inline src/appleseed/renderer/meta/tests/test_shaderparamparser.cpp]: Make expected value the first argument, this will lead to better messages if the assertion fails.

File: src/appleseed/renderer/meta/tests/test_shaderparamparser.cpp [hunk 2]
```diff
@@ -72,6 +72,24 @@ TEST_SUITE(Renderer_Modeling_ShaderParamParser)
...
+    }
+
+    TEST_CASE(ShaderParamParserColorArray)
+    {
+        ShaderParamParser parser("color[] 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0");
```
Comment (dictoon) [inline src/appleseed/renderer/meta/tests/test_shaderparamparser.cpp]: What happens if, say, the number of values in the array is not a multiple of 3? I think the edge cases deserve test cases, even more so than the main cases...

File: src/appleseed/renderer/modeling/shadergroup/shaderparam.h [hunk 3]
```diff
@@ -100,51 +105,76 @@ class APPLESEED_DLLSYMBOL ShaderParam
...
-        const float     vz);
+        const char*         name,
+        const float         vx,
+        const float         vy,
+        const float         vz);
+
+    // Create a vector array param.
+    static foundation::auto_release_ptr<ShaderParam> create_vector_array_param(
+        const char*         name,
+        std::vector<float>&         value);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/shadergroup/shaderparam.h]: Fix indentation.

File: src/appleseed/renderer/meta/tests/test_shaderparamparser.cpp (hunk 2 above)
Comment (est77) [inline src/appleseed/renderer/meta/tests/test_shaderparamparser.cpp]: In that case, an exception will be thrown.
